### PR TITLE
Deep copy views to avoid races

### PR DESCRIPTION
**Pull request by ian-mi**

### CODE REVIEW

File: metrics/resource_view.go [hunk 1]
```diff
@@ -295,6 +287,21 @@ func resourceFromKey(s string) *resource.Resource {
 	return r
 }
 
+func copyViews(views []*view.View) []*view.View {
+	viewsCopy := make([]*view.View, 0, len(resourceViews.views))
+	for _, v := range views {
+		c := *v
+		c.TagKeys = make([]tag.Key, 0, len(v.TagKeys))
```
Comment (vagababov) [inline metrics/resource_view.go:294]: use `copy` builtin.

File: metrics/resource_view.go [hunk 2]
```diff
@@ -295,6 +287,21 @@ func resourceFromKey(s string) *resource.Resource {
...
+		c.TagKeys = append(c.TagKeys, v.TagKeys...)
+		c.Aggregation = new(view.Aggregation)
+		*c.Aggregation = *v.Aggregation
```
Comment (vagababov) [inline metrics/resource_view.go:297]:
```suggestion
agg := *v.Aggregation
c.Aggregation = &agg
```

**Review by vagababov — COMMENTED** [review]
If only we could codegen deepcopy on it :)

File: metrics/resource_view.go (hunk 2 above)
Comment (ian-mi) [reply metrics/resource_view.go:297]: Isn't this equivalent?

Comment (vagababov) [reply metrics/resource_view.go:297]: Probably, but we're mostly doing this in the codebase. I.e. readability mostly.

File: metrics/resource_view.go [hunk 3]
```diff
@@ -295,6 +287,21 @@ func resourceFromKey(s string) *resource.Resource {
...
 }
 
+func copyViews(views []*view.View) []*view.View {
+	viewsCopy := make([]*view.View, 0, len(resourceViews.views))
+	for _, v := range views {
+		c := *v
+		c.TagKeys = make([]tag.Key, len(v.TagKeys))
+		copy(c.TagKeys, v.TagKeys)
+agg := *v.Aggregation
+c.Aggregation = &agg
```
Comment (mattmoor) [inline metrics/resource_view.go:297]: Format Go code:
```suggestion
		agg := *v.Aggregation
		c.Aggregation = &agg
```

**Review by mattmoor — COMMENTED** [review]
Produced via:
  `gofmt -s -w $(find -path './vendor' -prune -o -path './third_party' -prune -o -name '*.pb.go' -prune -o -type f -name '*.go' -print)`
  `goimports -w $(find -name '*.go' | grep -v vendor | grep -v third_party | grep -v .pb.go | grep -v wire_gen.go)`

**Comment by yanweiguo** [comment]
Are the failed tests relative?

**Comment by ian-mi** [comment]
The downstream test is failing because it does not always unregister views after each test and opencensus is no longer able to recognize that the aggregations are equal since it relied on a shallow comparison. https://github.com/knative/serving/pull/9497 should fix this.

**Pull request merged**